artichoke-backend/src/eval.rs

```diff
@@ -56,7 +57,13 @@ impl Eval for Artichoke {
         let context = Context::new(os_str_to_bytes(file.as_os_str())?.to_vec())
             .ok_or_else(|| ArgumentError::with_message("path name contains null byte"))?;
         self.push_context(context)?;
-        let code = self.read_source_file_contents(file)?.into_owned();
+        let code = self
+            .read_source_file_contents(file)
+            .map_err(|err| {
+                let message = format!("ruby: {} -- {}", err.message().as_bstr(), file.display());
```

Comment:
Sorry for not being clear on this. The underlying representation for a Ruby `String` and Artichoke exception messages is `Vec<u8>`. Because we are extracting the error message from an existing error, we end up with bytes, so we should stay with bytes.

Calling `as_bstr` on those bytes in the original message results in a transform if the message contains invalid UTF-8 byte sequences, which we'd like to avoid. Calling `display` on the path does the same destructive transform.

To help us stay in bytes, let's construct the new message like this (forgive the pseudocode, I'm on my mobile):

```rust
let mut message = b"ruby: ";
message.extend_from_slice(err.message());
message.extend_from_slice(b" - ");
message.extend_from_slice(scolapasta_path::os_str_to_bytes(path.into())?);
LoadError::from(message)
```

There is a fallible helper function in `scolapasta_path` which converts an `OsStr` to a byte slice and `Path` is freely convertible to `OsStr`. https://artichoke.github.io/artichoke/scolapasta_path/fn.os_str_to_bytes.html

Reply:
Oh I see. Thanks for the detailed explanation! I try to handle the `ConvertBytesError` the way [a similar case in `load.rs`](https://github.com/choznerol/artichoke/blob/c191d021bdd1f1953461ef99bbb1f824b8441f6b/artichoke-backend/src/load.rs#L101-L106) does.

Fixed in c191d02

Reply:
thank you!